File: frontend/src/utils/__tests__/repos.test.ts
import { describe, it, expect } from 'vitest';
import {
  validateRepoPattern,
  validateRepoPatterns,
  validateRepoFilterPatterns,
  filterRepos,
  parseFilterPatterns,
  serializeFilterPatterns,
} from '../repos';

describe('repos utility functions', () => {
  describe('validateRepoPattern', () => {
    it('accepts valid exact and wildcard patterns', () => {
      expect(validateRepoPattern('kubernetes/kubernetes')).toBeNull();
      expect(validateRepoPattern('kubernetes/*')).toBeNull();
      expect(validateRepoPattern('*/website')).toBeNull();
      expect(validateRepoPattern('*/*')).toBeNull();
      expect(validateRepoPattern('kubernetes/test-*')).toBeNull();
      expect(validateRepoPattern('k8s.io/release-1.28')).toBeNull();
      expect(validateRepoPattern('!kubernetes/steering')).toBeNull();
    });

    it('rejects empty or whitespace-only patterns', () => {
      expect(validateRepoPattern('')).toBe('Repository pattern cannot be empty');
      expect(validateRepoPattern('   ')).toBe('Repository pattern cannot be empty');
      expect(validateRepoPattern('!')).toBe('Repository pattern cannot be empty');
      expect(validateRepoPattern('!  ')).toBe('Repository pattern cannot be empty');
    });

    it('rejects patterns missing slash or with multiple slashes', () => {
      expect(validateRepoPattern('kubernetes')).toContain('must be in "owner/repo" format');
      expect(validateRepoPattern('!kubernetes')).toContain('must be in "owner/repo" format');
      expect(validateRepoPattern('a/b/c')).toContain('must be in "owner/repo" format');
      expect(validateRepoPattern('/kubernetes')).toContain('must be in "owner/repo" format');
      expect(validateRepoPattern('kubernetes/')).toContain('must be in "owner/repo" format');
    });

    it('rejects invalid characters', () => {
      expect(validateRepoPattern('kubernetes/foo$bar')).toContain('contains invalid character "$"');
      expect(validateRepoPattern('kubernetes / kubernetes')).toContain('contains invalid character " "');
    });

    it('rejects patterns exceeding 100 characters', () => {
      const longPattern = 'kubernetes/' + 'a'.repeat(95);
      expect(validateRepoPattern(longPattern)).toContain('exceeds maximum length of 100 characters');
    });
  });

  describe('validateRepoPatterns and validateRepoFilterPatterns', () => {
    it('validates a list of patterns', () => {
      expect(validateRepoPatterns(['kubernetes/*', 'golang/go', '!kubernetes/steering'])).toBeNull();
      expect(validateRepoPatterns(['', '  '])).toBeNull();
      expect(validateRepoPatterns(['valid/repo', 'invalid'])).toContain('must be in "owner/repo" format');
    });

    it('validates multiline text', () => {
      expect(validateRepoFilterPatterns('kubernetes/*\n!kubernetes/steering\ngolang/*')).toBeNull();
      expect(validateRepoFilterPatterns('kubernetes/*\n!\n')).toContain('cannot be empty');
      expect(validateRepoFilterPatterns('kubernetes/*\ninvalid_repo\n')).toContain('must be in "owner/repo" format');
    });
  });

  describe('parseFilterPatterns and serializeFilterPatterns', () => {
    it('parses mixed include and exclude lines', () => {
      const input = `
        kubernetes/*
        !kubernetes/steering
        golang/*
        !golang/proposal
      `;
      const parsed = parseFilterPatterns(input);
      expect(parsed.includes).toEqual(['kubernetes/*', 'golang/*']);
      expect(parsed.excludes).toEqual(['kubernetes/steering', 'golang/proposal']);
    });

    it('serializes includes and excludes to multi-line string', () => {
      const serialized = serializeFilterPatterns(
        ['kubernetes/*', 'golang/*'],
        ['kubernetes/steering', 'golang/proposal']
      );
      expect(serialized).toBe('kubernetes/*\ngolang/*\n!kubernetes/steering\n!golang/proposal');
    });
  });

  describe('filterRepos', () => {
    const repos = ['kubernetes/kubernetes', 'kubernetes/website', 'golang/go', 'kubernetes/test-infra'];

    it('returns all repos when patterns are empty', () => {
      expect(filterRepos(repos, [], [])).toEqual(repos);
    });

    it('filters repos by include wildcard pattern', () => {
      expect(filterRepos(repos, ['kubernetes/*'], [])).toEqual([
        'kubernetes/kubernetes',
        'kubernetes/website',
        'kubernetes/test-infra',
      ]);
    });

    it('filters repos by exclude wildcard pattern', () => {
      expect(filterRepos(repos, [], ['kubernetes/test-infra', 'golang/*'])).toEqual([
        'kubernetes/kubernetes',
        'kubernetes/website',
      ]);
    });

    it('filters repos by combined include and exclude patterns with exclude precedence', () => {
      expect(filterRepos(repos, ['kubernetes/*'], ['kubernetes/test-infra'])).toEqual([
        'kubernetes/kubernetes',
        'kubernetes/website',
      ]);
    });
  });
});

File: backend/internal/logic/repos_test.go
package logic

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	octodeckv1 "github.com/tallclair/octodeck/backend/internal/api/octodeck/v1"
	"github.com/tallclair/octodeck/backend/internal/config"
)

func TestValidateRepoPattern(t *testing.T) {
	tests := []struct {
		name    string
		pattern string
		wantErr bool
	}{
		{"valid exact repo", "kubernetes/kubernetes", false},
		{"valid wildcard repo", "kubernetes/*", false},
		{"valid wildcard org", "*/website", false},
		{"valid wildcard both", "*/*", false},
		{"valid prefix wildcard", "kubernetes/test-*", false},
		{"valid dots and hyphens", "k8s.io/release-1.28", false},
		{"valid exclude with exclamation", "!kubernetes/steering", false},
		{"empty string", "", true},
		{"only exclamation", "!", true},
		{"no slash", "kubernetes", true},
		{"too many slashes", "a/b/c", true},
		{"leading slash", "/kubernetes", true},
		{"trailing slash", "kubernetes/", true},
		{"invalid characters", "kubernetes/foo$bar", true},
		{"spaces inside", "kubernetes / kubernetes", true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := ValidateRepoPattern(tt.pattern)
			if tt.wantErr {
				assert.Error(t, err)
			} else {
				assert.NoError(t, err)
			}
		})
	}
}

func TestValidateRepoPatterns(t *testing.T) {
	require.NoError(t, ValidateRepoPatterns([]string{
		"kubernetes/*",
		"golang/go",
		"rust-lang/*",
		"!kubernetes/steering",
	}))
	require.NoError(t, ValidateRepoPatterns([]string{"", "  "})) // empty ignored
	require.Error(t, ValidateRepoPatterns([]string{"valid/repo", "invalid"}))
}

func TestFilterItemsByRepo(t *testing.T) {
	item1 := octodeckv1.Item_builder{Id: config.Ptr("1"), Repo: config.Ptr("kubernetes/kubernetes")}.Build()
	item2 := octodeckv1.Item_builder{Id: config.Ptr("2"), Repo: config.Ptr("kubernetes/website")}.Build()
	item3 := octodeckv1.Item_builder{Id: config.Ptr("3"), Repo: config.Ptr("golang/go")}.Build()
	item4 := octodeckv1.Item_builder{Id: config.Ptr("4"), Repo: config.Ptr("kubernetes/test-infra")}.Build()

	items := []*octodeckv1.Item{item1, item2, item3, item4}

	t.Run("empty patterns returns all items", func(t *testing.T) {
		res := FilterItemsByRepo(items, nil, nil)
		assert.Equal(t, items, res)
	})

	t.Run("include wildcard org keeps matching items", func(t *testing.T) {
		res := FilterItemsByRepo(items, []string{"kubernetes/*"}, nil)
		require.Len(t, res, 3)
		assert.Equal(t, "1", res[0].GetId())
		assert.Equal(t, "2", res[1].GetId())
		assert.Equal(t, "4", res[2].GetId())
	})

	t.Run("include multiple patterns", func(t *testing.T) {
		res := FilterItemsByRepo(items, []string{"kubernetes/website", "golang/*"}, nil)
		require.Len(t, res, 2)
		assert.Equal(t, "2", res[0].GetId())
		assert.Equal(t, "3", res[1].GetId())
	})

	t.Run("exclude wildcard pattern removes matching items", func(t *testing.T) {
		res := FilterItemsByRepo(items, nil, []string{"kubernetes/test-*", "golang/*"})
		require.Len(t, res, 2)
		assert.Equal(t, "1", res[0].GetId())
		assert.Equal(t, "2", res[1].GetId())
	})

	t.Run("combined include and exclude patterns with exclude precedence", func(t *testing.T) {
		// Include all kubernetes repos EXCEPT test-infra
		res := FilterItemsByRepo(items, []string{"kubernetes/*"}, []string{"kubernetes/test-infra"})
		require.Len(t, res, 2)
		assert.Equal(t, "1", res[0].GetId())
		assert.Equal(t, "2", res[1].GetId())
	})
}

func TestFilterRepos(t *testing.T) {
	repos := []string{"kubernetes/kubernetes", "kubernetes/website", "golang/go", "kubernetes/test-infra"}

	t.Run("include filter", func(t *testing.T) {
		res := FilterRepos(repos, []string{"kubernetes/*"}, nil)
		assert.Equal(t, []string{"kubernetes/kubernetes", "kubernetes/website", "kubernetes/test-infra"}, res)
	})

	t.Run("exclude filter", func(t *testing.T) {
		res := FilterRepos(repos, nil, []string{"kubernetes/test-infra", "golang/*"})
		assert.Equal(t, []string{"kubernetes/kubernetes", "kubernetes/website"}, res)
	})

	t.Run("combined include and exclude filter", func(t *testing.T) {
		res := FilterRepos(repos, []string{"kubernetes/*", "golang/*"}, []string{"kubernetes/test-infra", "golang/go"})
		assert.Equal(t, []string{"kubernetes/kubernetes", "kubernetes/website"}, res)
	})
}
